[WebGPU] Support creating WebGPU Texture backed by IOSurface

tuankiet65 opened the pull request:
#### b5f8e33ec9c453271a02ead84582fa3e2dd3c76e
<pre>
[WebGPU] Support creating WebGPU Texture backed by IOSurface
<a href="https://bugs.webkit.org/show_bug.cgi?id=242197">https://bugs.webkit.org/show_bug.cgi?id=242197</a>
&lt;rdar://96612079&gt;

Reviewed by Myles C. Maxfield.

On Apple platforms, WebGL content is composited into the web view by rendering
into an IOSurface, then providing that IOSurface to the compositor. We&apos;d like to
do the same with WebGPU. In WebGPU, content is rendered into render attachments, which
themselves are textures. This patch adds support for creating WebGPU textures backed
by an IOSurface, so that when contents are rendered into a texture, the actual pixel
data is stored inside the IOSurface, which can then be passed to the compositor.
This is straightforward to implement because WebGPU textures are Metal textures under
the hood, and Metal provides official API to create IOSurface-backed textures.

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj: Link to IOSurface.framework.
* Source/WebGPU/WebGPU/Device.h: Add parameter to createTexture() to optionally take
a IOSurfaceRef. Declare new method to validate IOSurface-backed texture descriptors.
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Device::validateCreateIOSurfaceBackedTexture): Add method to validate
IOSurface-backed texture descriptors.
(WebGPU::storageMode): On Mac / Mac Catalyst, force storage mode to MTLStorageModeManaged
when the texture is backed by an IOSurface. This is required by the Metal driver.
(WebGPU::Device::createTexture): Add logic to validate and create IOSurface-backed texture.

Canonical link: <a href="https://commits.webkit.org/252674@main">https://commits.webkit.org/252674@main</a>
</pre>

litherum commented on Source/WebGPU/WebGPU/Texture.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#import "WebGPU.h"
```
This shouldn't be necessary. All .cpp files must `#include "config.h"` as their first include, and that file has `#include "WebGPU.h"` inside it.

tuankiet65 replied:
Yeah that I know, my editor is angry about the missing definitions though (which yeah, it will be there when the header file gets included into .cpp files)

litherum commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -1961,20 +1961,32 @@ static MTLPixelFormat pixelFormat(WGPUTextureFormat textureFormat)
+static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, bool supportsNonPrivateDepthStencilTextures, bool isBackedByIOSurface)
 {
+    // IOSurface-backed texture must be MTLStorageModeManaged, else the driver will assert with
+    // "Texture Descriptor Validation IOSurface textures must use MTLStorageModeManaged"
```
I think this second clause is unnecessary. I'd rewrite the comment to say "Metal validation requires IOSurface-backed textures to be MTLStorageModeManaged."

Is this comment true for devices with unified memory? If not, we should probably change the logic in this function.

litherum replied:
(See below comment about `isForDepthOrStencil`) Can we add something here like

```
ASSERT(!isForDepthOrStencil || !isBackedByIOSurface);
```

tuankiet65 replied:
> Is this comment true for devices with unified memory? If not, we should probably change the logic in this function.

I hit the assertion on a M1 MacBook Pro so yes

litherum commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -1961,20 +1961,32 @@ static MTLPixelFormat pixelFormat(WGPUTextureFormat textureFormat)
+        return MTLStorageModeShared;
+#endif
+
     if (!supportsNonPrivateDepthStencilTextures)
```
I know you didn't write this, but I think this logic is wrong. This condition should only be true if the texture we're trying to create is a depth or stencil texture. Can we add in another argument into this function

```
bool isForDepthOrStencil
```

and then change this condition to be

```
if (isForDepthOrStencil && !supportsNonPrivateDepthStencilTextures)
```

tuankiet65 replied:
Maybe for another patch? I don't like that we're doing two things in a commit

litherum replied:
ok. can you at least add a FIXME then?

tuankiet65 replied:
I'll add in a FIXME (even though I'm planning to send in the patch right after this is merged)

litherum commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -1961,20 +1961,32 @@ static MTLPixelFormat pixelFormat(WGPUTextureFormat textureFormat)
+#else
+        // MTLStorageModeManaged is not supported on iOS.
+        return MTLStorageModeShared;
+#endif
```
We have a block down below that has this code duplicated. Can we deduplicate it somehow?

I'm also not so sure about the usefulness of the comment.

tuankiet65 replied:
I like to document stuff that I don't know (because I only know that when compiling for the iOS simulator). But I can remove it.

tuankiet65 replied:
I tried looking into deduplicating the code but I found that it makes the logic hard to read. I think the current logic is the most straightforward - if it's backed by an IOSurface then we just make it Managed/Shared.

litherum commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -2051,12 +2063,17 @@ static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, bool supportsNonP
-    id<MTLTexture> texture = [m_device newTextureWithDescriptor:textureDescriptor];
+    id<MTLTexture> texture = nil;
+    if (ioSurfaceBacking)
+        texture = [m_device newTextureWithDescriptor:textureDescriptor iosurface:ioSurfaceBacking plane:0];
```
How confident are you that there aren't additional validation rules that are necessary to check before trying to run this line of code? What happens if the texture descriptor indicates we should be trying to make a 3D texture? Or a mipmapped texture? Or a multisample texture?

I don't think this patch needs to handle all these cases, but we at least need a placeholder in the source and a bug filed to determine this stuff.

tuankiet65 replied:
Not very confident - I'll look into what kind of textures can be backed by an `IOSurface` and check for it.

tuankiet65 replied:
I added some validation.

litherum commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -1339,6 +1339,46 @@ static bool textureViewFormatCompatible(WGPUTextureFormat format1, WGPUTextureFo
     return true;
 }
 
+bool Device::validateCreateIOSurfaceBackedTexture(const WGPUTextureDescriptor& descriptor, const Vector<WGPUTextureFormat>& viewFormats, IOSurfaceRef backing)
```
Are we confident that the implementation of this function is comprehensive?

tuankiet65 replied:
I'm _fairly_ confident since I strictly followed what Metal people told me about.

tuankiet65 replied:
Now to think about it, I'd like to test more thing in the `IOSurfaceRef`

tuankiet65 replied:
I added two more checks: pixel format and non-planar.

tuankiet65 committed:
[WebGPU] Support creating WebGPU Texture backed by IOSurface
https://bugs.webkit.org/show_bug.cgi?id=242197
<rdar://96612079>

Reviewed by Myles C. Maxfield.

On Apple platforms, WebGL content is composited into the web view by rendering
into an IOSurface, then providing that IOSurface to the compositor. We'd like to
do the same with WebGPU. In WebGPU, content is rendered into render attachments, which
themselves are textures. This patch adds support for creating WebGPU textures backed
by an IOSurface, so that when contents are rendered into a texture, the actual pixel
data is stored inside the IOSurface, which can then be passed to the compositor.
This is straightforward to implement because WebGPU textures are Metal textures under
the hood, and Metal provides official API to create IOSurface-backed textures.

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj: Link to IOSurface.framework.
* Source/WebGPU/WebGPU/Device.h: Add parameter to createTexture() to optionally take
a IOSurfaceRef. Declare new method to validate IOSurface-backed texture descriptors.
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Device::validateCreateIOSurfaceBackedTexture): Add method to validate
IOSurface-backed texture descriptors.
(WebGPU::storageMode): On Mac / Mac Catalyst, force storage mode to MTLStorageModeManaged
when the texture is backed by an IOSurface. This is required by the Metal driver.
(WebGPU::Device::createTexture): Add logic to validate and create IOSurface-backed texture.

Canonical link: https://commits.webkit.org/252674@main
